s2a_e2e_test.go

```diff
@@ -405,3 +410,101 @@ func TestNewTLSClientConfigFactoryWithoutTokenManager(t *testing.T) {
 		t.Errorf("tls.Config has unexpected certificate: got: %v, want: %v", got, want)
 	}
 }
+
+// startHTTPServer runs an HTTP server on a random local port and serves a /hello endpoint.
+func startHTTPServer(t *testing.T) string {
+	cert, _ := tls.X509KeyPair(serverCertpem, serverKeypem)
+	tlsConfig := tls.Config{
+		Certificates: []tls.Certificate{cert},
+	}
+	s := http.NewServeMux()
+	s.HandleFunc("/hello", func(w http.ResponseWriter, req *http.Request) {
+		fmt.Fprintf(w, "hello")
+	})
+	lis, err := tls.Listen("tcp", ":0", &tlsConfig)
+	if err != nil {
+		t.Errorf("net.Listen(tcp, :0) failed: %v", err)
+	}
+	go func() {
+		http.Serve(lis, s)
+	}()
+	return lis.Addr().String()
+}
+
+// runHTTPClient starts an HTTP client and talks to an HTTP server using S2A.
+func runHTTPClient(t *testing.T, clientS2AAddress, serverAddr string, fallbackOpts *FallbackOptions) {
+	dialTLSContext := NewS2ADialTLSContextFunc(&ClientOptions{
+		S2AAddress:   clientS2AAddress,
+		EnableV2:     true,
+		FallbackOpts: fallbackOpts,
+	})
+
+	tr := http.Transport{
+		DialTLSContext: dialTLSContext,
+	}
+
+	client := &http.Client{Transport: &tr}
+	reqURL := fmt.Sprintf("https://%s/hello", serverAddr)
+	t.Logf("reqURL is set to: %v", reqURL)
+	req, err := http.NewRequest(http.MethodGet, reqURL, nil)
+	if err != nil {
+		t.Errorf("error creating new HTTP request: %v", err)
+	}
+	resp, err := client.Do(req)
+	if err != nil {
+		t.Errorf("error making client HTTP request: %v", err)
+	}
+	respBody, err := io.ReadAll(resp.Body)
+	if err != nil {
+		t.Errorf("error reading HTTP response: %v", err)
+	}
+	if got, want := string(respBody), "hello"; got != want {
+		t.Errorf("expecting HTTP response:[%s], got [%s]", want, got)
+	}
+}
+func TestHTTPEndToEndUsingFakeS2AOverTCP(t *testing.T) {
```

Comment:
Can we also test the gRPC APIs at this layer?

Reply:
done!